netket/nn/linear.py

```diff
@@ -0,0 +1,417 @@
+        and after each spatial dimension.
+      input_dilation: `None`, or a sequence of `n` integers, giving the
+        dilation factor to apply in each spatial dimension of `inputs`.
+        Convolution with input dilation `d` is equivalent to transposed
+        convolution with stride `d`.
+      kernel_dilation: `None`, or a sequence of `n` integers, giving the
+        dilation factor to apply in each spatial dimension of the convolution
+        kernel. Convolution with kernel dilation is also known as 'atrous
+        convolution'.
+      feature_group_count: integer, default 1. If specified divides the input
+        features into groups.
+      use_bias: whether to add a bias to the output (default: True).
+      dtype: the dtype of the computation (default: float32).
+      precision: numerical precision of the computation see `jax.lax.Precision`
+        for details.
+      kernel_init: initializer for the convolutional kernel.
+      bias_init: initializer for the bias.
+    """
+
+    features: int
+    kernel_size: Union[int, Iterable[int]]
+    strides: Optional[Iterable[int]] = None
+    padding: Union[str, Iterable[Tuple[int, int]]] = "SAME"
+    input_dilation: Optional[Iterable[int]] = None
+    kernel_dilation: Optional[Iterable[int]] = None
+    feature_group_count: int = 1
+    use_bias: bool = True
+    bias_dtype: Any = None
+    dtype: Dtype = jnp.float32
+    bias_dtype: Optional[Dtype] = None
+    precision: Any = None
+    kernel_init: Callable[[PRNGKey, Shape, Dtype], Array] = default_kernel_init
+    bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = zeros
+
+    def setup(self):
+        if self.bias_dtype is None:
+            self.bias_dtype = self.dtype
+
+    @compact
+    def __call__(self, inputs: Array) -> Array:
+        """Applies a convolution to the inputs.
+        Args:
+          inputs: input data with dimensions (batch, spatial_dims..., features).
+        Returns:
+          The convolved data.
+        """
+        dtype = nkjax.maybe_promote_to_complex(self.dtype, inputs.dtype)
+
+        inputs = jnp.asarray(inputs, dtype)
+
+        if isinstance(self.kernel_size, int):
+            kernel_size = (self.kernel_size,)
+        else:
+            kernel_size = self.kernel_size
+
+        is_single_input = False
+        if inputs.ndim == len(kernel_size) + 1:
+            is_single_input = True
+            inputs = jnp.expand_dims(inputs, axis=0)
+
+        strides = self.strides or (1,) * (inputs.ndim - 2)
+
+        in_features = inputs.shape[-1]
+        assert in_features % self.feature_group_count == 0
+        kernel_shape = kernel_size + (
+            in_features // self.feature_group_count,
+            self.features,
+        )
+        kernel = self.param("kernel", self.kernel_init, kernel_shape, self.dtype)
+        kernel = jnp.asarray(kernel, dtype)
+
+        dimension_numbers = flax.linen.linear._conv_dimension_numbers(inputs.shape)
```

Comment:
This line throws `NameError: name 'flax' is not defined`.

Reply:
thanks. it's resolved now.